### PR TITLE
Fix types + cjs/mjs imports

### PR DESCRIPTION
**Pull request by miunau**
Re: #8

- Removes default exports (https://basarat.gitbook.io/typescript/main-1/defaultisbad)
- Explicit exports (no `export * from './foo.js'`) (I can't find the link right now, but tldr it's bad for ASTs and tree shaking)
- Fixed cjs types (added `types` to package.json)
- Separate mjs and cjs types to `.d.ts` and `.d.cts` so they resolve correctly (ATTW reports no problems now)
- Bumped to 2.0.0 for breaking API change (removing default exports)

### CODE REVIEW

**Comment by miunau** [comment]
By the way, is there a reason you've configured esbuild here? I'm unsure if it's even needed here to be honest, if it were my project I'd probably try to make do with two tsconfig files instead and lose the big dep :)

**Comment by miunau** [comment]
ping @corrideat could you take a look?

**Comment by corrideat** [comment]
Hi @minau! Sorry that it took a while to respond.

I've taken a look and I have a few comments.

> Removes default exports (https://basarat.gitbook.io/typescript/main-1/defaultisbad)
> Bumped to 2.0.0 for breaking API change (removing default exports)

This is perhaps my bigger concern, that it breaks the API. Unless there's something actively breaking, I'm more open to adding additional `exports` (in the `package.json`) for specific exports. This doesn't break the API _and_ ensures that you only get the parts you need / want, even if tree-shaking fails for some reason.

> Explicit exports (no export * from './foo.js') (I can't find the link right now, but tldr it's bad for ASTs and tree shaking)

Related to the above. Do you have an example of something breaking from the default exports in the current form? 

> Fixed cjs types (added types to package.json)

This looks good.

> Separate mjs and cjs types to .d.ts and .d.cts so they resolve correctly (ATTW reports no problems now)

This may indeed be needed. I wonder if there's a better way to do this (other than copying files) but I don't think there is (other than using `moduleResolution: bundler`, which also cannot be forced upon users).

TL;DR: The parts related to types look fine, the parts about exports, I'm less sure about.

**Comment by corrideat** [comment]
Hi again,

I was doing some more testing over here: <https://github.com/Exact-Realty/ts-rfc8188>.

1) Regarding default imports, they work but are reported as potentially not working by ATTW
2) Regarding the `cts` files, it seems like simple copying and renaming won't work, because those files reference other files and then `tsc` fails. For example,

```
// foo.d.ts -> foo.d.cts
export { baz } from '../bar.js';
```

The issue is that `tsc` will complain that `bar.js` is an ESM module and compilation will fail. So, it seems like the contents also need to be rewritten so that instead of `.js` it says `.cjs` (and of course, replacing this with a regexp works but it could break). It's quite unfortunate that that's the state of affairs.

---
Now, I was planning on making a new release of this package anyhow. Is there something specific that's broken with the default exports? Otherwise, I think I'll make a new release fixing the types and perhaps we can revisit breaking changes later on. I'll try to also add exports for specific parts, like in the other package I mentioned, so that there are fewer things that can go wrong when bundling.

**Comment by miunau** [comment]
Sorry I was offline for a few days! Yes, I see the fixed types now, thank you!
I guess you can close this PR or pick whatever you like from it :) I understand not wanting to break the default export, that's fair!
